src/databricks/labs/ucx/azure/locations.py

```diff
@@ -0,0 +1,170 @@
+                if not credential.read_only:
+                    app_id_mapping_write[credential.azure_service_principal.application_id] = credential.name
+                    continue
+                if credential.read_only:
+                    app_id_mapping_read[credential.azure_service_principal.application_id] = credential.name
+            # if managed identity based credential, fetch the application_id of the managed identity
+            if credential.azure_managed_identity:
+                application_id = self._azurerm.managed_identity_client_id(
+                    credential.azure_managed_identity.access_connector_id,
+                    credential.azure_managed_identity.managed_identity_id,
+                )
+                if not application_id:
+                    continue
+                if not credential.read_only:
+                    app_id_mapping_write[application_id] = credential.name
+                    continue
+                if credential.read_only:
+                    app_id_mapping_read[application_id] = credential.name
+
+        return app_id_mapping_write, app_id_mapping_read
+
+    def _prefix_credential_name_mapping(self) -> tuple[dict[str, str], dict[str, str]]:
+        # get managed identity/service principal's application id to storage credential name mapping
+        # for all non read-only and read-only credentials
+        app_id_mapping_write, app_id_mapping_read = self._app_id_credential_name_mapping()
+
+        # use the application id to storage credential name mapping to create prefix to storage credential name mapping
+        prefix_mapping_write = {}
+        prefix_mapping_read = {}
+        for permission_mapping in self._resource_permissions.load():
+            if permission_mapping.client_id in app_id_mapping_write:
+                prefix_mapping_write[permission_mapping.prefix] = app_id_mapping_write[permission_mapping.client_id]
+                continue
+            if permission_mapping.client_id in app_id_mapping_read:
+                prefix_mapping_read[permission_mapping.prefix] = app_id_mapping_read[permission_mapping.client_id]
+        return prefix_mapping_write, prefix_mapping_read
+
+    def _create_location_name(self, location_url: str) -> str:
+        # generate the UC external location name
+        before_at, _, after_at = location_url.partition('@')
+        container_name = before_at.removeprefix("abfss://")
+        res_name = after_at.replace(".dfs.core.windows.net", "").rstrip("/").replace("/", "_")
+        return f"{container_name}_{res_name}"
+
+    def _create_external_location(
+        self, location_url: str, prefix_mapping_write: dict[str, str], prefix_mapping_read: dict[str, str]
+    ) -> str | None:
+        location_name = self._create_location_name(location_url)
+
+        # get container url as the prefix
+        parsed_url = urlparse(location_url)
+        container_url = f"{parsed_url.scheme}://{parsed_url.netloc}/"
```

Comment:
This is loosing data - what if the prefix contains a path?

Reply:
`access.py` crawls [container to principal mapping](https://github.com/databrickslabs/ucx/blob/e0da29aee2c10e91f01b1c315812fa0c3e3eadf9/src/databricks/labs/ucx/azure/access.py#L79), so for a given location we extract the container url and use it as the key to find the principal and hence the storage credential to be used.